source/Engine.h

```diff
@@ -159,7 +161,13 @@ class Engine {
 
 	void CalculateStep();
 
-	void MoveShip(const std::shared_ptr<Ship> &ship);
+	// Thread-safe wrapper for MoveShip.
+	using ShipResourceProvider = ResourceProvider<std::list<Visual>, std::list<std::shared_ptr<Flotsam>>,
```

Comment:
are you using this class ResourceProvider anywhere else? Do you plan to? If not, we can make this a much more simple object that does not need to be template mayhem

Reply:
It's going to be used in more places. Visuals will get a provider and BatchDrawList will also have a provider.

Reply:
in that case its good